libs/utils/src/filter.c

```diff
@@ -463,367 +555,327 @@ static celix_status_t celix_filter_compile(celix_filter_t* filter) {
+        filter1->operand == CELIX_FILTER_OPERAND_NOT) {
+        assert(filter1->children != NULL);
+        assert(filter2->children != NULL);
+        int sizeSrc = celix_arrayList_size(filter1->children);
+        int sizeDest = celix_arrayList_size(filter2->children);
+        if (sizeSrc == sizeDest) {
+            int i;
+            int k;
+            int sameCount = 0;
+            for (i = 0; i < sizeSrc; ++i) {
+                bool same = false;
+                celix_filter_t* srcPart = celix_arrayList_get(filter1->children, i);
+                for (k = 0; k < sizeDest; ++k) {
+                    celix_filter_t* destPart = celix_arrayList_get(filter2->children, k);
+                    filter_match_filter(srcPart, destPart, &same);
+                    if (same) {
+                        sameCount += 1;
+                        break;
                     }
                 }
-                result = sameCount == sizeSrc;
-            }
-        } else { //compare attr and value
-            bool attrSame = false;
-            bool valSame = false;
-            if (filter1->attribute == NULL && filter2->attribute == NULL) {
-                attrSame = true;
-            } else if (filter1->attribute != NULL && filter2->attribute != NULL) {
-                attrSame = celix_utils_stringEquals(filter1->attribute, filter2->attribute);
             }
+            return sameCount == sizeSrc;
+        }
+        return false;
+    }
 
-            if (filter1->value == NULL  && filter2->value == NULL) {
-                valSame = true;
-            } else if (filter1->value != NULL && filter2->value != NULL) {
-                valSame = celix_utils_stringEquals(filter1->value, filter2->value);
-            }
+    // compare attr and value
+    bool attrSame = false;
+    bool valSame = false;
+    if (filter1->attribute == NULL && filter2->attribute == NULL) {
+        attrSame = true;
+    } else if (filter1->attribute != NULL && filter2->attribute != NULL) {
+        attrSame = celix_utils_stringEquals(filter1->attribute, filter2->attribute);
+    }
 
-            result = attrSame && valSame;
-        }
+    if (filter1->value == NULL && filter2->value == NULL) {
+        valSame = true;
+    } else if (filter1->value != NULL && filter2->value != NULL) {
+        valSame = celix_utils_stringEquals(filter1->value, filter2->value);
     }
 
-    return result;
+    return attrSame && valSame;
 }
 
-const char* celix_filter_getFilterString(const celix_filter_t *filter) {
+const char* celix_filter_getFilterString(const celix_filter_t* filter) {
     if (filter != NULL) {
         return filter->filterStr;
     }
     return NULL;
 }
 
-const char* celix_filter_findAttribute(const celix_filter_t *filter, const char *attribute) {
-    const char *result = NULL;
+const char* celix_filter_findAttribute(const celix_filter_t* filter, const char* attribute) {
```

Comment:
This interface seems not well-defined: for filter `(|(a=b)(a=c))`, what should `celix_filter_findAttribute` returns for `a`?

Reply:
Updated doxygen to clarify this.